File: uhrwerk-common/src/main/java/io/qimia/uhrwerk/common/model/HashKeyUtils.java
package io.qimia.uhrwerk.common.model;

import net.openhft.hashing.LongHashFunction;

public class HashKeyUtils {

  public static Long targetKey(TargetModel target) {
    assert target.tableId != null : "Target.tableId can't be null";
    assert target.connectionId != null : "Target.connectionId can't be null";
    assert target.format != null && !target.format.isEmpty()
        : "Target.format can't be null or empty";

    return hashKey(
        new StringBuilder()
            .append(target.tableId)
            .append(target.connectionId)
            .append(target.format));
  }

  public static Long tableKey(TableModel table) {
    assert table.area != null && !table.area.isEmpty() : "Table.area can't be null or empty";
    assert table.vertical != null && !table.vertical.isEmpty()
        : "Table.vertical can't be null or empty";
    assert table.name != null && !table.name.isEmpty() : "Table.name can't be null or empty";
    assert table.version != null && !table.version.isEmpty()
        : "Table.version can't be null or empty";

    return hashKey(
        new StringBuilder()
            .append(table.area)
            .append(table.vertical)
            .append(table.name)
            .append(table.version));
  }

  public static Long tableKey(DependencyModel dependency) {
    assert dependency.area != null && !dependency.area.isEmpty()
        : "Dependency.area can't be null or empty";
    assert dependency.vertical != null && !dependency.vertical.isEmpty()
        : "Dependency.vertical can't be null or empty";
    assert dependency.tableName != null && !dependency.tableName.isEmpty()
        : "Dependency.tableName can't be null or empty";
    assert dependency.version != null && !dependency.version.isEmpty()
        : "Dependency.version can't be null or empty";

    return hashKey(
        new StringBuilder()
            .append(dependency.area)
            .append(dependency.vertical)
            .append(dependency.tableName)
            .append(dependency.version));
  }

  public static Long sourceKey(SourceModel source) {
    assert source.tableId != null : "Source.tableId can't be null";
    assert source.connectionId != null : "Source.connectionId can't be null";
    assert source.path != null && !source.path.isEmpty() : "Source.path can't be null or empty";
    assert source.format != null && !source.format.isEmpty()
        : "Source.format can't be null or empty";
    return hashKey(
        new StringBuilder()
            .append(source.tableId)
            .append(source.connectionId)
            .append(source.path)
            .append(source.format));
  }

  public static Long dependencyKey(DependencyModel dependency) {
    assert dependency.tableId != null : "Dependency.tableId can't be null";
    assert dependency.dependencyTargetId != null : "Dependency.dependencyTargetId can't be null";
    assert dependency.dependencyTableId != null : "Dependency.dependencyTableId can't be null";
    return hashKey(
        new StringBuilder()
            .append(dependency.tableId)
            .append(dependency.dependencyTargetId)
            .append(dependency.dependencyTableId));
  }

  public static Long connectionKey(ConnectionModel connection) {
    assert connection.name != null && !connection.name.isEmpty()
        : "Connection.name can't be null or empty";
    return hashKey(new StringBuilder().append(connection.name));
  }

  private static Long hashKey(StringBuilder strBuilder) {
    return LongHashFunction.xx().hashChars(strBuilder);
  }
}

File: uhrwerk-metastore/src/main/kotlin/io/qimia/uhrwerk/common/metastore/model/HashKeyUtils.kt
package io.qimia.uhrwerk.common.metastore.model

import io.qimia.uhrwerk.common.model.TargetModel
import net.openhft.hashing.LongHashFunction

object HashKeyUtils {

    fun tableKey(table: TableModel): Long {
        assert(!table.area.isNullOrEmpty()) { "Table.area can't be null or empty" }
        assert(!table.vertical.isNullOrEmpty()) { "Table.vertical can't be null or empty" }
        assert(!table.name.isNullOrEmpty()) { "Table.name can't be null or empty" }
        assert(!table.version.isNullOrEmpty()) { "Table.version can't be null or empty" }
        return hashKey(
            StringBuilder()
                .append(table.area)
                .append(table.vertical)
                .append(table.name)
                .append(table.version)
        )
    }

    fun tableKey(dependency: DependencyModel): Long {
        assert(!dependency.area.isNullOrEmpty()) { "Dependency.area can't be null or empty" }
        assert(!dependency.vertical.isNullOrEmpty()) { "Dependency.vertical can't be null or empty" }
        assert(!dependency.tableName.isNullOrEmpty()) { "Dependency.tableName can't be null or empty" }
        assert(!dependency.version.isNullOrEmpty()) { "Dependency.version can't be null or empty" }
        return hashKey(
            StringBuilder()
                .append(dependency.area)
                .append(dependency.vertical)
                .append(dependency.tableName)
                .append(dependency.version)
        )
    }

    fun targetKey(target: TargetModel): Long {
        assert(target.tableKey != null) { "Target.tableKey can't be null" }
        assert(target.connectionKey != null) { "Target.connectionKey can't be null" }
        assert(!target.format.isNullOrEmpty()) { "Target.format can't be null or empty" }
        return hashKey(
            StringBuilder()
                .append(target.tableKey)
                .append(target.connectionKey)
                .append(target.format)
        )
    }


    fun tableKey(area: String?, vertical: String?, tableName: String?, version: String?): Long {
        assert(!area.isNullOrEmpty()) { "area can't be null or empty" }
        assert(!vertical.isNullOrEmpty()) { "vertical can't be null or empty" }
        assert(!tableName.isNullOrEmpty()) { "tableName can't be null or empty" }
        assert(!version.isNullOrEmpty()) { "version can't be null or empty" }
        return hashKey(
            StringBuilder()
                .append(area)
                .append(vertical)
                .append(tableName)
                .append(version)
        )
    }

    fun sourceKey(source: SourceModel2): Long {
        assert(source.tableKey != null) { "Source.tableKey can't be null" }
        assert(source.connectionKey != null) { "Source.connectionKey can't be null" }
        assert(!source.path.isNullOrEmpty()) { "Source.path can't be null or empty" }
        assert(!source.format.isNullOrEmpty()) { "Source.format can't be null or empty" }
        return hashKey(
            StringBuilder()
                .append(source.tableKey)
                .append(source.connectionKey)
                .append(source.path)
                .append(source.format)
        )
    }

    fun dependencyKey(dependency: DependencyModel): Long {
        assert(dependency.tableKey != null) { "Dependency.tableKey can't be null" }
        assert(dependency.dependencyTargetKey != null) { "Dependency.dependencyTargetKey can't be null" }
        assert(dependency.dependencyTableKey != null) { "Dependency.dependencyTableKey can't be null" }
        return hashKey(
            StringBuilder()
                .append(dependency.tableKey)
                .append(dependency.dependencyTargetKey)
                .append(dependency.dependencyTableKey)
        )
    }

    fun connectionKey(connection: ConnectionModel): Long {
        assert(!connection.name.isNullOrEmpty()) { "Connection.name can't be null or empty" }
        return hashKey(StringBuilder().append(connection.name))
    }

    fun secretKey(secret: SecretModel): Long {
        assert(secret.name != null && secret.name!!.isNotEmpty()) { "Secret.name can't be null or empty" }
        return hashKey(StringBuilder().append(secret.name))
    }

    fun secretKey(name: String): Long {
        assert(name != null && name!!.isNotEmpty()) { "Secret.name can't be null or empty" }
        return hashKey(StringBuilder().append(name))
    }


    private fun hashKey(strBuilder: StringBuilder): Long {
        return LongHashFunction.xx128low().hashChars(strBuilder)
    }
}